### ISSUE TITLE
Slow response on bufenter (winenter?)

### ISSUE THREAD

**Issue by jondkinney** [killphi/vim-legend#11]
Hi there!

I really enjoy Cadre and vim-legend, however, I'm having some performance issues with it. Using terminal vim 7.4 in iTerm 2 on OS X Yosemite inside of tmux, when I toggle signs initially they show up very quickly, and if I leave a split there isn't any delay, but trying to re-enter a split with visible coverage signs takes 4-5 seconds before vim becomes responsive again. It also seems to sort of 'flash' as it re-draws that split with the coverage info. This only happens when the vim-legend signs/cadre signs are present. It also seems to be slow to toggle the signs off regardless of moving in/out of a split. Any ideas what might be causing that or how I can troubleshoot it? cc: @nyarly 

Thanks!

**Comment by nyarly** [comment]
So glad you're enjoying the tools!

There's two things this makes me think might be happening: 

1) there are some reasonably aggressive autocmds in vim-legend right now, such that it might reload the signs information twice.

2) When it does start a reload, it does this in a very simple way which may do more work than it needs to do.

Let me ask you: how big are the files you're editing? I rarely have more than a quarter second - the flash is noticeable but not a 4 second process.

**Comment by jondkinney** [comment]
@nyarly thanks for the response! I'm not sure if my issue is fully fixed, but the last time I was using cadre/vim-legend it was very snappy (I don't remember if it was on the same project or not). Hopefully I just needed to reboot my machine or something. I don't remember what fixed it (if anything) but I'll keep a look out for slowness in the future and try to help debug further if necessary. The little flash you mentioned is what I'm seeing now and is totally tolerable!

I just wanted to mention that I did see good performance last time I used it so it's not a system wide thing.

Thanks again!

**Comment by killphi** [comment]
Sorry for the long delay, I was away for a while.

First of all, thanks for being a user and even more thanks for reporting issues!

As soon as I manage to have gotten back into my stuff, I'll have a look at this and will have a go at measuring and improvements.
Won't make any promises though, especially since I don't use OSX.

Greetings

**Comment by nyarly** [comment]
@jondkinney sorry for the late fix here - I recently ran into cases where I was having exactly the same problem you were having, and a few minutes with the vim profiler and a google later I found that fugitive has the same problem. Apparently using signs can be incredibly slow, and the solution is to minimize setting and removing them. I just merged a fix that compares updates to the coverage signs to the existing ones and changes them only as they actually change. Let me know if that works better for you.
